Migrate template to nf-core-utils plugin

edmundmiller opened the pull request:
<!--
Many thanks for contributing to nf-core/tools!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a release.

Learn more about contributing: https://github.com/nf-core/tools/tree/main/.github/CONTRIBUTING.md
-->

## PR checklist

- [ ] This comment contains a description of changes (with reason)
- [ ] `CHANGELOG.md` is updated
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] Documentation in `docs` is updated

edmundmiller committed:
feat(template): migrate to nf-core-utils plugin

Integrate nf-core-utils plugin v0.3.0 to replace utility subworkflows
with centralized plugin functions, following RFC proposal #51.

Major changes:
- Add nf-core-utils@0.3.0 plugin to nextflow.config
- Replace UTILS_NEXTFLOW_PIPELINE subworkflow with direct plugin calls
- Replace UTILS_NFCORE_PIPELINE subworkflow with plugin functions
- Replace UTILS_NFSCHEMA_PLUGIN subworkflow with nf-schema plugin calls
- Remove entire subworkflows/nf-core/ directory
- Update modules.json to remove nf-core subworkflows

Functions now provided by nf-core-utils plugin:
- getWorkflowVersion, dumpParametersToJSON, checkCondaChannels
- completionEmail, completionSummary, imNotification
- checkConfigProvided, checkProfileProvided
- processVersionsFromFile, workflowVersionToChannel

Functions retained in local subworkflow:
- paramsSummaryMultiqc, softwareVersionsToYAML (uses plugin functions)
- getGenomeAttribute, validateInputSamplesheet
- toolCitationText, methodsDescriptionText

Closes: #3685
Related: nf-core/proposals#51

feat(template): migrate to nf-core-utils plugin

- Add nf-core-utils@0.3.0 plugin to nextflow.config
- Replace all nf-core subworkflow imports with plugin imports
  - utils_nextflow_pipeline functions -> direct plugin calls
  - utils_nfcore_pipeline functions -> plugin/nf-core-utils
  - utils_nfschema_plugin functions -> plugin/nf-schema
- Use plugin functions for version processing:
  - processVersionsFromFile() instead of local implementation
  - workflowVersionToChannel() instead of local implementation
- Remove entire subworkflows/nf-core/ directory
- Update modules.json: empty nf-core subworkflows section
- Update test expectations for removed subworkflows

Closes #3685

mashehu commented on nf_core/pipeline-template/subworkflows/local/utils_nfcore_pipeline_pipeline/main.nf:
```diff
@@ -227,6 +193,55 @@ workflow PIPELINE_COMPLETION {
+        })
+}
+
+//
+// Get workflow summary for MultiQC
+//
+def paramsSummaryMultiqc(summary_params) {
```
wrap this whole function inside a `{% if multiqc %`

mashehu commented on nf_core/pipeline-template/main.nf:
```diff
@@ -23,6 +23,20 @@ include { PIPELINE_INITIALISATION } from './subworkflows/local/utils_nfcore_{{ s
+include { getWorkflowVersion      } from 'plugin/nf-core-utils'
+include { dumpParametersToJSON    } from 'plugin/nf-core-utils'
+include { checkCondaChannels      } from 'plugin/nf-core-utils'
+include { completionEmail         } from 'plugin/nf-core-utils'
```
needs a `{% if email %}` wrapper

mashehu commented on nf_core/pipeline-template/nextflow.config:
```diff
@@ -313,20 +308,49 @@ manifest {
 plugins {
-    id 'nf-schema@2.5.1' // Validation of pipeline parameters and creation of an input channel from a sample sheet
+    id 'nf-schema@2.4.2' // Validation of pipeline parameters and creation of an input channel from a sample sheet
+    id 'nf-core-utils@0.3.0' // Utility functions for nf-core pipelines
```
```suggestion
    id 'nf-core-utils@1.0.0' // Utility functions for nf-core pipelines
```

mashehu commented on tests/subworkflows/test_lint.py:
```diff
@@ -20,10 +19,10 @@ def test_subworkflows_lint(self):
-        self.subworkflow_remove.remove("utils_nfschema_plugin", force=True)
+        """Test linting a pipeline with no nf-core subworkflows installed.
+
+        The template now has no nf-core subworkflows by default (migrated to nf-core-utils plugin).
```
```suggestion
```

adamrtalbot committed:
feat(template): conditionally include completionEmail import

Add Jinja2 conditional to only include completionEmail function
from nf-core-utils plugin when email functionality is enabled.

adamrtalbot committed:
docs(tests): simplify test_subworkflows_lint_empty docstring

Remove implementation details from docstring.

adamrtalbot committed:
feat(template): conditionally include MultiQC functions

Add Jinja2 conditional to only include paramsSummaryMultiqc function
when MultiQC is enabled in the pipeline.

adamrtalbot committed:
fix(modules.json): Ignore modules.json formatting which broke jinja templating

mirpedrol commented on nf_core/pipeline-template/subworkflows/local/utils_nfcore_pipeline_pipeline/main.nf:
```diff
@@ -213,6 +197,57 @@ workflow PIPELINE_COMPLETION {
+//
+// Get channel of software versions used in pipeline in YAML format
+//
+def softwareVersionsToYAML(ch_versions) {
```
should this function be also added to the plugin?
